### PR TITLE
[WasdTV] Add extractor

### LINKED ISSUE #3041: Support for wasd.tv?
### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.03.08.1**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

Europe

### Example URLs

https://wasd.tv/StasSatori?record=919135
https://wasd.tv/spacemita/videos
https://wasd.tv/spacemita/videos?record=907755
https://wasd.tv/spacemita/clips
https://wasd.tv/spacemita/clips?clip=26804

### Description

Hi.
It would be nice to add wasd.tv support.
For now it looks like only some patches for youtube-dl doing it.

https://gitea.datahoarding.agency/ZenulAbidin/youtube-dl/commit/cdf98a34eae6a6d31f4b2c5a9ef36cace4cb3d92
https://copr.fedorainfracloud.org/coprs/vr5/wasd.tv/

### Verbose log

```shell
./yt-dlp -vU "https://wasd.tv/StasSatori?record=919135"
[debug] Command-line config: ['-vU', 'https://wasd.tv/StasSatori?record=919135']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, err utf-8, pref UTF-8
[debug] yt-dlp version 2022.03.08.1 [c0c2c57d3] (zip)
[debug] Python version 3.10.2 (CPython 64bit) - Linux-5.16.12-200.fc35.x86_64-x86_64-with-glibc2.34
[debug] exe versions: ffmpeg 4.4.1 (setts), ffprobe 4.4.1
[debug] Optional libraries: Crypto, sqlite
[debug] Proxy map: {}
Latest version: 2022.03.08.1, Current version: 2022.03.08.1
yt-dlp is up to date (2022.03.08.1)
[debug] [generic] Extracting URL: https://wasd.tv/StasSatori?record=919135
[generic] StasSatori?record=919135: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] StasSatori?record=919135: Downloading webpage
[generic] StasSatori?record=919135: Extracting information
[debug] Looking for video embeds
ERROR: Unsupported URL: https://wasd.tv/StasSatori?record=919135
Traceback (most recent call last):
  File "/home/anleal/./yt-dlp/yt_dlp/YoutubeDL.py", line 1389, in wrapper
    return func(self, *args, **kwargs)
  File "/home/anleal/./yt-dlp/yt_dlp/YoutubeDL.py", line 1459, in __extract_info
    ie_result = ie.extract(url)
  File "/home/anleal/./yt-dlp/yt_dlp/extractor/common.py", line 617, in extract
    ie_result = self._real_extract(url)
  File "/home/anleal/./yt-dlp/yt_dlp/extractor/generic.py", line 3985, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://wasd.tv/StasSatori?record=919135
```

### PR DESCRIPTION
**Pull request by hatienl0i261299**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Close #3041

### CODE REVIEW

File: yt_dlp/extractor/wasdtv.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    urljoin,
+    try_get,
+)
+
+
+class WASDTVBaseIE(InfoExtractor):
+    _API_BASE = 'https://wasd.tv/api/'
+    _THUMBNAIL_SIZES = ('small', 'medium', 'large')
+
+    def _fetch(self, *args, **kwargs):
```
Comment (Ashish0804) [inline yt_dlp/extractor/wasdtv.py:22]: why is this needed?
u are not even using fata=false
the code will never show ur ExtractorError message.

File: yt_dlp/extractor/wasdtv.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    int_or_none,
+    parse_iso8601,
+    urljoin,
+    try_get,
+)
+
+
+class WASDTVBaseIE(InfoExtractor):
+    _API_BASE = 'https://wasd.tv/api/'
+    _THUMBNAIL_SIZES = ('small', 'medium', 'large')
```
Comment (Ashish0804) [inline yt_dlp/extractor/wasdtv.py:20]: only used once, no  need for it to be class var

File: yt_dlp/extractor/wasdtv.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+            if not thumbnail_url:
+                continue
+            thumbnails.append({
+                'url': thumbnail_url,
+                'preference': index,
+            })
+        return thumbnails
+
+    def _extract_og_title(self, url, item_id):
+        return self._og_search_title(self._download_webpage(url, item_id))
```
Comment (Ashish0804) [inline yt_dlp/extractor/wasdtv.py:50]: y?

Comment (Ashish0804) [reply yt_dlp/extractor/wasdtv.py:50]: Useless webpage download call.

File: yt_dlp/extractor/wasdtv.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+    def _real_extract(self, url):
+        container = self._get_container(url)
+        stream = try_get(container, lambda x: x['media_container_streams'][0])
+        media = try_get(stream, lambda x: x['stream_media'][0])
+        media_meta = media.get('media_meta')
```
Comment (Ashish0804) [inline yt_dlp/extractor/wasdtv.py:56]: when these are None, u will get error upon using .get
If they are required, make them fatal else do or {}

File: yt_dlp/extractor/wasdtv.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        media_url, is_live = self._get_media_url(media_meta)
+        video_id = media.get('media_id') or container.get('media_container_id')
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(media_url, video_id, 'mp4')
+        self._sort_formats(formats)
+        return {
+            'id': compat_str(video_id),
```
Comment (Ashish0804) [inline yt_dlp/extractor/wasdtv.py:62]:
```suggestion
            'id': str(video_id),
```

**Comment by Ashish0804** [comment]
I didn't review further as i see same mistakes( proly not the right word but whatever) in the rest of the code aswell.
Modify them accordingly

File: yt_dlp/extractor/wasdtv.py (hunk 1 above)
Comment (Ashish0804) [reply yt_dlp/extractor/wasdtv.py:22]: I see that u are using it to clean up the code.
Yeah it's nice to have this then.
Just hardcode the args u are consuming and are def needed and rest can be left to kwargs and args

File: yt_dlp/extractor/wasdtv.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+class WASDTVBaseIE(InfoExtractor):
+    _API_BASE = 'https://wasd.tv/api/'
+
+    def _fetch(self, path, video_id, description, query={}):
+        response = self._download_json(
+            urljoin(self._API_BASE, path),
+            video_id=video_id,
+            note=f'Downloading {description} metadata',
+            errnote=f'Unable to download {description} metadata',
+            query=query)
```
Comment (pukkandan) [inline yt_dlp/extractor/wasdtv.py:23]:
```suggestion
    def _fetch(self, path, video_id, description, query={}):
        response = self._download_json(
            f'https://wasd.tv/api/{path}', video_id, query=query,
            note=f'Downloading {description} metadata',
            errnote=f'Unable to download {description} metadata')
```

File: yt_dlp/extractor/wasdtv.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        thumbnails = []
+        for index, thumbnail_size in enumerate(['small', 'medium', 'large']):
+            thumbnail_url = thumbnails_dict.get(thumbnail_size)
+            if not thumbnail_url:
+                continue
+            thumbnails.append({
+                'url': thumbnail_url,
+                'preference': index,
+            })
+        return thumbnails
```
Comment (pukkandan) [inline yt_dlp/extractor/wasdtv.py:41]:
```suggestion
        return [{
            'url': url,
            'preference': index,
        } for index, url in enumerate(
            traverse_obj(thumbnails_dict, (('small', 'medium', 'large'),))) if url]
```

File: yt_dlp/extractor/wasdtv.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+    def _real_extract(self, url):
+        container = self._get_container(url)
+        stream = try_get(container, lambda x: x['media_container_streams'][0])
```
Comment (pukkandan) [inline yt_dlp/extractor/wasdtv.py:45]: Not a amndatory change, but fyi, this can be simplified to:
```suggestion
        stream = traverse_obj(container, ('media_container_streams', 0))
```

File: yt_dlp/extractor/wasdtv.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+    def _real_extract(self, url):
+        clip_id = self._match_id(url)
+        clip = self._fetch(f'v2/clips/{clip_id}', video_id=clip_id, description='clip')
+        clip_data = clip.get('clip_data')
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(clip_data.get('url'), video_id=clip_id, ext='mp4')
+        self._sort_formats(formats)
+        return {
+            'id': clip_id,
+            'title': clip.get('clip_title') or self._og_search_title(self._download_webpage(url, clip_id)),
```
Comment (pukkandan) [inline yt_dlp/extractor/wasdtv.py:163]:
```suggestion
            'title': clip.get('clip_title') or self._og_search_title(
                self._download_webpage(url, clip_id, fatal=False)),
```

File: yt_dlp/extractor/wasdtv.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+            'title': container.get('media_container_name') or self._og_search_title(self._download_webpage(url, video_id)),
+            'description': container.get('media_container_description'),
+            'thumbnails': self._extract_thumbnails(media_meta.get('media_preview_images')),
+            'timestamp': parse_iso8601(container.get('created_at')),
+            'view_count': int_or_none(stream.get('stream_current_viewers' if is_live else 'stream_total_viewers')),
+            'is_live': is_live,
+            'formats': formats,
+            'subtitles': subtitles,
+        }
+
```
Comment (pukkandan) [inline yt_dlp/extractor/wasdtv.py:65]:
```suggestion
    def _get_container(self, url):
        raise NotImplementedError('Must be defined in subclasses')

    def _get_media_url(self, media_meta):
        raise NotImplementedError('Must be defined in subclasses')

```
for self-documentation

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/wasdtv.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+            'view_count': int,
+        },
+    }]
+
+    def _get_container(self, url):
+        container_id = self._match_id(url)
+        return self._fetch(
+            f'v2/media-containers/{container_id}',
+            video_id=container_id,
+            description='media container')
```
Comment (pukkandan) [inline yt_dlp/extractor/wasdtv.py:125]:
```suggestion
        return self._fetch(
            f'v2/media-containers/{container_id}', container_id, 'media container')
```

File: yt_dlp/extractor/wasdtv.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+                'media_container_status': 'RUNNING',
+            },
+            video_id=channel_id,
+            description='running media containers')
+        if not containers:
+            raise ExtractorError(f'{nickname} is offline', expected=True)
+        return containers[0]
+
+    def _get_media_url(self, media_meta):
+        return media_meta.get('media_url'), True
```
Comment (pukkandan) [inline yt_dlp/extractor/wasdtv.py:98]:
```suggestion
        return media_meta['media_url'], True
```

File: yt_dlp/extractor/wasdtv.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+                'media_container_type': 'SINGLE',
+                'media_container_status': 'RUNNING',
+            },
+            video_id=channel_id,
+            description='running media containers')
+        if not containers:
+            raise ExtractorError(f'{nickname} is offline', expected=True)
+        return containers[0]
+
+    def _get_media_url(self, media_meta):
```
Comment (pukkandan) [inline yt_dlp/extractor/wasdtv.py:97]: Does this have `media_archive_url`? If not, `_get_media_url` is actually same for both the classes

File: yt_dlp/extractor/wasdtv.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+        channel_id = channel.get('channel_id')
+        containers = self._fetch(
+            'v2/media-containers',
+            query={
+                'channel_id': channel_id,
+                'media_container_type': 'SINGLE',
+                'media_container_status': 'RUNNING',
+            },
+            video_id=channel_id,
+            description='running media containers')
```
Comment (pukkandan) [inline yt_dlp/extractor/wasdtv.py:92]:
```suggestion
        containers = self._fetch(
            'v2/media-containers', channel_id, 'running media containers',
            query={
                'channel_id': channel_id,
                'media_container_type': 'SINGLE',
                'media_container_status': 'RUNNING',
            })
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/wasdtv.py (hunk 13 above)
Comment (hatienl0i261299) [reply yt_dlp/extractor/wasdtv.py:95]: Yes, sure, It has media_archive_url

**Comment by pukkandan** [comment]
> - [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

??

**Comment by hatienl0i261299** [comment]
> > * [x]  I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)
> 
> ??

As you can see, in the description of issue https://github.com/yt-dlp/yt-dlp/issues/3041 has a link `youtube-dl` that extracted this site, and the author of issue want add this site to yt-dlp, so that i copied and enhanced this extractor to yt-dlp.

**Comment by pukkandan** [comment]
this? https://github.com/ytdl-org/youtube-dl/pull/26500

**Comment by hatienl0i261299** [comment]
> this? [ytdl-org/youtube-dl#26500](https://github.com/ytdl-org/youtube-dl/pull/26500)

Im not sure, i used this link, and for now this link cant access
image

But maybe your link is true.

**Comment by pukkandan** [comment]
then how can I know the code is unlicense?

> it is in public domain or released under Unlicense

> provide reliable evidence

**Comment by pukkandan** [comment]
If you can confirm your code is based on that PR, that is sufficient evidence. Otherwise, the original author will need to confirm

**Comment by hatienl0i261299** [comment]
> If you can confirm your code is based on that PR, that is sufficient evidence. Otherwise, the original author will need to confirm

Oh sure, i understood what you mean, i apologize about it.

**Comment by Lesmiscore** [comment]
Looks like the LICENSE file is modified about a year ago, although I can't confirm its content.
https://web.archive.org/web/20210508010911/https://gitea.datahoarding.agency/ZenulAbidin/youtube-dl

**Comment by mesb1** [comment]
Hi.
Not sure i clearly understand discussion here related to license and unavailable repo (it was available when i created ticket).
Is there anything i can help you with?

**Comment by pukkandan** [comment]
This PR is using code from that (now unavailable) repo according to OP. But we are not sure how the said code is licensed. Until it can be confirmed to be in the public domain, this can't be merged

**Comment by mesb1** [comment]
> Until it can be confirmed to be in the public domain, this can't be merged

It looks like author is also exists on github:
https://github.com/ZenulAbidin

is it enough for you if he confirm it?

**Comment by pukkandan** [comment]
assuming they are the original author, yes

**Comment by mesb1** [comment]
Hi.

Author has moved it's repo to gitlab.
So now this commit available here:

https://gitlab.com/ZenulAbidin/youtube-dl/-/commit/cdf98a34eae6a6d31f4b2c5a9ef36cace4cb3d92

**Comment by pukkandan** [comment]
The commit appears to be the same as https://github.com/ytdl-org/youtube-dl/pull/26500
